core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala

```diff
@@ -0,0 +1,207 @@
+          producer.commitTransaction()
+          numMessagesProcessed += records.size
+        }
+        iteration += 1
+      } catch {
+        case e : Exception =>
+          fail (s"Unexpected exception occurred while copying messages from $inputTopic to $outputTopic", e)
+      }
+    }
+
+    scheduler.shutdown()
+
+    val verifyingConsumer = createConsumerAndSubscribeToTopics("randoGroup", List(outputTopic), readCommitted = true)
+    val outputRecords = TestUtils.pollUntilAtLeastNumRecords(verifyingConsumer, numInputRecords).map { case(record) =>
+      assertCommittedAndGetValue(record)
+    }
+    val recordSet = outputRecords.toSet
+    assertEquals(numInputRecords, recordSet.size)
+
+    val expectedValues = (0 until numInputRecords).toSet
+    assertEquals(s"Missing messages: ${expectedValues -- recordSet}", expectedValues, recordSet)
+
+    consumer.close()
+    producer.close()
+    verifyingConsumer.close()
+  }
+
+  private def createConsumerAndSubscribeToTopics(groupId: String, topics: List[String], readCommitted: Boolean = false) = {
+    val props = new Properties()
+    if (readCommitted)
+      props.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed")
+    props.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "200")
+    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
+    props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "10000")
+    props.put(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, "3000")
+    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+
+    val consumer = TestUtils.createNewConsumer(TestUtils.getBrokerListStrFromServers(servers), groupId = groupId,
+      securityProtocol = SecurityProtocol.PLAINTEXT, props = Some(props))
+    consumer.subscribe(topics)
+    consumer
+  }
+
+  private def createTopics() =  {
+    val topicConfig = new Properties()
+    topicConfig.put(KafkaConfig.MinInSyncReplicasProp, 2.toString)
+    TestUtils.createTopic(zkUtils, inputTopic, numPartitions, numServers, servers, topicConfig)
+    TestUtils.createTopic(zkUtils, outputTopic, numPartitions, numServers, servers, topicConfig)
+  }
+
+    // Verifies that the record was intended to be committed by checking the suffix of the value. If true, this
+  // will return the value with the '-committed' suffix removed.
+  private def assertCommittedAndGetValue(record: ConsumerRecord[Array[Byte], Array[Byte]]) : Int = {
+    val recordValue = new String(record.value(), "UTF-8")
+    assertTrue(recordValue.endsWith("committed"))
+    recordValue.replace("-committed", "").toInt
+  }
+
+  private def recordValue(record: ConsumerRecord[Array[Byte], Array[Byte]]) : Int = {
+    val recordValue = new String(record.value(), "UTF-8")
+    recordValue.replace("-committed", "").replace("-aborted", "").toInt
+  }
+
+  private def producerRecord(topic: String, consumerRecord: ConsumerRecord[Array[Byte], Array[Byte]], willBeCommitted: Boolean) = {
```

Comment:
Maybe instead of all the string munging, we could use the message key (or dare I suggest, a header?) to indicate whether the record should be committed or aborted.